src/hooks/useModal/index.tsx

```diff
@@ -0,0 +1,74 @@
+import React, { useState, useCallback, ReactElement, createContext, useMemo, useContext } from 'react';
+import { createPortal } from 'react-dom';
+import { match } from 'ts-pattern';
+
+import { noop } from '../../utils/utils';
+import Modal, { type ModalProps } from '../../ui/Modal';
+import { MODAL_ROOT } from './ModalRoot';
+
+export type OpenGlobalModalProps = {
+  modalProps: ModalProps;
+  childElement: ReactElement;
+};
+export interface GlobalModalProviderProps {
+  children: ReactElement;
+}
+export interface GlobalModalContextInterface {
+  openModal: (props: OpenGlobalModalProps) => void;
+}
+
+const GlobalModalContext = createContext<GlobalModalContextInterface>({
+  openModal: noop,
+});
+
+export const GlobalModalProvider = ({ children }: GlobalModalProviderProps) => {
+  // Idea from https://dev.to/nurislamov/simple-modals-queue-in-react-4g6c
+  const [globalModalQueue, setGlobalModalQueue] = useState<Array<OpenGlobalModalProps>>([]);
+
+  const openModal = useCallback((props: OpenGlobalModalProps) => {
+    setGlobalModalQueue((currentQue) => [...currentQue, props]);
```

Comment:
just curious, why is this better than
```
setGlobalModalQueue([... globalModalQueue, props]);
```
?

Reply:
It's to make use of Functional Updates.

There are several advantages to this approach:
1. Safe Updates: By using functional updates, you avoid potential issues related to relying on the current state (count) directly for updates. This helps prevent unexpected behavior in cases of asynchronous updates or consecutive updates.
2. Optimization: React can optimize component rendering when you use functional updates. React keeps track of the previous state values between renders, and when you use functional updates, it can avoid unnecessary re-renders, thus improving performance.
3. Readability: Functional updates use functions, which can make complex state update logic more explicit and easier to trace and debug.

Reply:
Nice explanation, thanks! Definitely good reasoning especially 1 and 2.